kotlin/nucleotide-count/src/main/kotlin/Dna.kt

```diff
@@ -0,0 +1,20 @@
+import java.lang.IllegalArgumentException
+
+class Dna(sequence: String) {
+
+    private val dnaSequence: String = sequence
+    private val validNucleotides = listOf('A', 'C', 'G', 'T')
+
+    init {
+        if (!dnaSequence.all { it in validNucleotides } && dnaSequence.isNotEmpty()) {
```

Comment:
Kotlin has a dedicated method to check if requirements are met. `require` throws IAE when the check returns false.
Also the regex check might be a bit better fit for this check.

``` 
   private val nucleotideRegex = "[CTAG]*".toRegex()
    init {
        require(nucleotideRegex.matches(sequence))
    }
```